feat(forms): change methods of Validators to arrow functions

xiaoxiangmoe opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [x] Other... Please describe:


## What is the current behavior? What is the new behavior?

TypeScript has a rule [@typescript-eslint/unbound-method](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/unbound-method.md) and angular has [Built-in validator functions](https://angular.io/guide/form-validation#built-in-validator-functions).

We use validator function as below:

```ts
new FormControl(this.hero.name, [
    Validators.required,
    Validators.minLength(4),
])
```

let's see the types of  Validators.required and Validators.minLength.

```ts
class Validators {
    static required(control: AbstractControl): ValidationErrors | null;
    static minLength(minLength: number): ValidatorFn;
}
```

![image](https://user-images.githubusercontent.com/8111351/101648087-6e88ab80-3a74-11eb-802a-06f6ca188ff6.png)
![image](https://user-images.githubusercontent.com/8111351/101648296-a263d100-3a74-11eb-807d-1384bf6db64c.png)

The `Validators.required` is a *class method* and the return type of `Validators.minLength` is a function `ValidatorFn`.

Eslint think a method may contain `this`, and if someone use a direct reference to the method, such as `Validators.required`, the `this`, which in the implementation of the method body, may point to an incorrect object, like `window`/`undefined` or something else.

And eslint think if a function contain this, it will told us by [this parameters](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters). So we don't have to worry about `this` value pointing problem of `ValidatorFn` and eslint will not blame.

---

Current behavior：
![image](https://user-images.githubusercontent.com/8111351/101649975-6af62400-3a76-11eb-8964-9d3ce303fc00.png)

New behavior: 
![image](https://user-images.githubusercontent.com/8111351/101650120-924cf100-3a76-11eb-9f11-88167499071a.png)


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

It would be better if we can change `class Validators` to `namespace Validators`.

xiaoxiangmoe committed:
feat(forms): change methods of Validators to arrow functions

petebacondarwin commented:
The rule in eslint has an option to ignore static class methods: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/unbound-method.md#ignorestatic

I think that this is the appropriate solution here.

AndrewKushnir commented:
Closing this PR based on the comment from Pete. Thank you.

xiaoxiangmoe commented:
@petebacondarwin @AndrewKushnir
static method can also use this to call another method, and we use it in our codebase. So we can't use ignore static.

petebacondarwin commented:
> static method can also use this.someStaticMethod() to call another static method

Why not just use the class name? If you are relying upon `this` to allow static methods to be, in some way, polymorphic, then, to be honest, that looks like a suspicious use of `this` to me.

I don't believe that the framework should have to be modified to enable this rule in this way.

xiaoxiangmoe commented:
@petebacondarwin
That also looks like a suspicious use of this to me, and I also hate this very much, and I will never use it.
But I can't rely on other library don't use it. 

---

We once had a bug:
A library provided by another team in our company used this syntax, but when I refactored it, I changed `function foo() { console.log('some log'); return ClassName.someStaticMethod(); }` to `const foo =  ClassName.someStaticMethod` and it works well.

But one day we update the library, it caused a serious failure. because someStaticMethod call `this.somePrivateStaticMethod` in method body. But TypeScript and eslint don't find this.

---

After that, we open `@typescript-eslint/unbound-method`. 

If some library suggests that we can use the static method as a function, we'll replace it with static fields with arrow function, which sends the message to the outside world that you can refer to it without any dependency on dynamic `this`

petebacondarwin commented:
I can see your problem but I think that Angular is not responsible for poor libraries. If you want you could have your own re-export of the `Validators` class as a namespace. You could then turn off this rule just for that file.

xiaoxiangmoe commented:
@petebacondarwin 
Can I create a pr to change  `class Validators` into `namespace Validators`

xiaoxiangmoe commented:
> Angular is not responsible for poor libraries.

I think the usage of using static method as function is poor. (Using `this` in static method is poor, also.)
So I enable the eslint rule and create this PR)

petebacondarwin commented:
Let me discuss it further with @AndrewKushnir

xiaoxiangmoe commented:
@petebacondarwin Anything update?

petebacondarwin commented:
@AndrewKushnir did some investigation into the G3 code base and found this was breaking a number of projects. I am not sure of the details. But it might mean that the effort to solve this for you is not worth the cost. I'll leave @AndrewKushnir  to update with more details.

AndrewKushnir commented:
@xiaoxiangmoe, while testing this change in Google's codebase we came across an issue (which is likely related to TypeScript itself, not Angular), see https://github.com/microsoft/TypeScript/issues/41977. We can not perform further tests until the mentioned issue is resolved. Once we're unblocked, we can run another set of tests to see what's the level of effort required to land it. I will reopen this PR for now (will mark it as blocked) and we'll review it again once we're unblocked. Thank you.

alxhub commented:
Closing as we can't land this internally due to the TS bug, which is still pending.
